keps/sig-network/20190125-ingress-api-group.md

```diff
@@ -389,38 +382,100 @@ type IngressSpec struct {
   Class string
   ...
 }
+```
+
+##### Interoperability with previous annotation
+
+The `kubernetes.io/ingress.class` annotation on v1beta1 Ingresses will be used
+to set the value of the `class` field on v1 Ingresses as part of API conversion.
+When both the `class` field and annotation are set, the field will take
+priority. The controller MAY emit a warning event if the user sets conflicting
+(different) values for the annotation and field.
 
-...
+#### IngressClass Resource
 
-// IngressClass represents the class of the Ingress, referenced by the
-// ingress.spec. IngressClass will be a non-namespaced Cluster resource.
+Additionally, we propose adding an `IngressClass` resource to provide additional
+data about the Ingress class. This resource will be non-namespaced. The
+IngressClass resource is an optional way to provide additional configuration for
+a specific class of Ingress. This allows us to evolve the API to express
+concepts such as levels of service associated with a given Ingress controller.
+
+The name of this IngressClass resource will be tied to any Ingresses with the
+same value for the class field or annotation. An IngressClass resource can exist
+without any Ingresses referencing it, and an Ingress can have a class value that
```

Comment:
What does it mean to have an Ingress class field with a value that is not the name of an IngressClass ?  why is that allowable?

Reply:
As it is implemented now, the value from the annotation is set on the field through conversion. So an annotation on a v1beta1 ingress will be translated to a v1 class field (and vice versa on the way back). Assuming that implementation makes sense, there would be no way to ensure the existence of the corresponding IngressClass resource since the annotation existed before the resource. If we tried to require a corresponding IngressClass resource and disconnected the field and the annotation, I think that cross resource validation could get tricky/messy to implement.

Reply:
I see.  This seems problematic to evolve.

Assume I am running 2 ingress controllers that understand the annotation, where the values "foo" and "bar" mean the different controllers should handle it.

I have written v1b1 instances that use that annotation.

Now I create an IngressClass instance named "foo", which uses "bar" as the driver.  Don't ask why I would do this, average users don't understand our API evolution as well as you do.

Obviously annotations "foo" are not the same as field "foo".

If I upgrade the "foo" controller to a version that understands v1, any ingress instance that set "foo" annotation in v1b1 -> "foo" field in v1 -> "bar" controller.

Might it be better to say:
* No conversion of annotation <-> field
* Some way of defaulting the field when not specified
* Both annotation and field are immutable
* The field is not defaulted when the annotation is set
* The field and annotation are mutually exclusive on create
* The annotation is deprecated in v1

Controllers that did not support the annotation are not required to support it.

Controllers that support v1 must support IngressClass indirection

To round-trip with v1b1 we need a DIFFERENT annotation or an equivalent field in v1b1.

Reply:
@thockin Thanks for the clarification, this makes a lot of sense. I'm on board with separating the class field and annotation. A few follow up questions on the specifics of implementation though:

1. Why would we need a default for the Ingress class field? It seems like it would be difficult to get this right, especially if this field needed to represent an existing IngressClass resource. Could we make this field optional instead?
2. Why would the annotation and field need to be immutable? As I understand it, the annotation is currently mutable, it seems like it would be difficult to transition to that being immutable. I don't mind considering the field immutable, would this be primarily to avoid complexity for consumers of this API?

Reply:
> Why would we need a default for the Ingress class field

What does it mean if you create an Ingress with no class?  Which controller should take it?  Now, maybe you have only one controller, so you don't care about class, but how do you tell the controller that?  The general answer is "handle any ingress whose class points to a class whose driver is you".    If there's no class field provided, what do we do?  We'll have to think thru the transition here.

The way StorageClass works is that there's a label on one StorageClass which indicates "I am 
 the default".  An admission controller sets that class on any PVC that doesn't have a class specified.  If there are no StorageClasses (or none set as default) it does nothing, which amounts to what we have today with Ingress, right?

what does it mean to be mutable?  De-provision from one and provision the other?  I was trying to cut complexity, but if you think that's not feasible, OK..  Now you have to define the "both" case :)

Reply:
@thockin Thanks for the great feedback! You're right, this is more problematic than I'd initially realized. Agree on most of this, including: 

- No conversion of annotation <-> field
- The field is not defaulted when the annotation is set
- The field and annotation are mutually exclusive on create
- The annotation is deprecated in v1

I think the following are problematic though:

1. Some way of defaulting the field when not specified. 
Unfortunately, I don't think it will be possible to set a good default value here. We simply can't guarantee that any IngressClass resource will exist in a cluster, so even if we were able to indicate that an IngressClass resource should be treated as default, that wouldn't allow us to guarantee that the class field would always have a value. We need to support backwards compatibility with all the existing v1beta1 Ingresses and they are not guaranteed to have any Ingress class (annotation or field). I've talked through different potential approaches here with @bowei and we can't think of a good way to ensure that there's always a value set for Ingress class _and_ maintain backwards compatibility. I do think it could be worthwhile to encourage implementations to provide optional defaulting webhooks for Ingresses if end users want a specific implementation to be the default in their cluster, but I don't think any approach to defaulting would be able to cover all potential use cases here. 

2. Both annotation and field are immutable.
The annotation has been mutable up to this point, I don't think we can add more restrictive validation at this point. Since the field is new, we could potentially make it immutable, but I think that may result in a poor user experience. I can imagine Ingress users starting their use of Ingress without specifying a class and then realizing that they should have added a specific class. Or maybe they made a typo in their class, or want to switch to a different Ingress implementation. Allowing the value of the field to change matches the pattern that has already been established with the annotation and likely provides a simpler, kubectl apply friendly, user experience. If we were to make this field immutable, I don't think it simplifies much for implementations. At the end of the day, implementations will still need to handle create/update/delete events, and they'll still need to be able to spin down any resources associated with an ingress. Given that implementations have already will still need to support Ingress class annotations changing, I don't think it's unreasonable for the field to behave similarly.

As long as that all sounds reasonable, I'll work on getting this PR updated to reflect this discussion.

Reply:
@thockin I just pushed an update that reflects all of the above, including a default value for IngressClass following the StorageClass pattern. I'd initially interpreted that point as requiring a class to be specified for each Ingress which seemed impossible, but using this to simply set a sane default when applicable makes a lot of sense. Feel free to disregard my reasoning in the first point above.

Reply:
@robscott and I talked through this - followup will be coming.

Reply:
Just pushed an update that:
- Clarifies what will happen when class is not specified as a field or annotation. 
- Updates the class field to be considered immutable.
- Clarifies what will happen when multiple IngressClass resources have the default annotation applied

Reply:
Just pushed one more point of clarification that the Ingress class field will be added to previous API versions to ensure that it can be safely round tripped.